authorization/invitation/service/invitation_service.go

```diff
@@ -0,0 +1,154 @@
+		}
+	}
+
+	// If we didn't successfully find a valid resource already, it means the inviteTo is a UUID
+	if inviteToResource == nil {
+		// Attempt to lookup the identity of the organization, team or security group that invitations will be issued for
+		inviteToIdentity, err = s.repo.Identities().Load(ctx, inviteToUUID)
+		if err != nil {
+			// That didn't work, try to lookup a resource with the same ID value
+			inviteToResource, err = s.repo.ResourceRepository().Load(ctx, inviteTo)
+			if err != nil {
+				return errors.NewNotFoundError(fmt.Sprintf("invalid identifier '%s' provided for organization, team, security group or resource", inviteTo), inviteTo)
+			}
+		}
+	}
+
+	if inviteToIdentity != nil {
+		// Load the resource for the identity
+		identityResource, err = s.repo.ResourceRepository().Load(ctx, *inviteToIdentity.IdentityResourceID)
+		if err != nil {
+			return errors.NewInternalErrorFromString(ctx, "error loading resource for identity")
+		}
+
+		// Confirm that the issuing user has the necessary scope to manage members for the organization, team or security group
+		scope, err := s.permSvc.HasScope(ctx, issuingUserId, *inviteToIdentity.IdentityResourceID, authorization.ManageMembersScope)
+		if err != nil {
+			return errors.NewInternalError(ctx, err)
+		}
+
+		if !scope {
+			return errors.NewForbiddenError(fmt.Sprintf("user requires %s scope to invite other users", authorization.ManageMembersScope))
+		}
+
+		// We only allow membership in some identity types - confirm that we are inviting to an organization, team or security group
+		if !authorization.CanHaveMembers(identityResource.ResourceType.Name) {
+			return errors.NewInternalErrorFromString(ctx, "may only invite a user to an organization, team or security group")
+		}
+	} else if inviteToResource != nil {
+		// Confirm that the issuing user has the manage members scope for the resource
+		scope, err := s.permSvc.HasScope(ctx, issuingUserId, inviteToResource.ResourceID, authorization.ManageMembersScope)
+		if err != nil {
+			return errors.NewInternalError(ctx, err)
+		}
+
+		if !scope {
+			return errors.NewForbiddenError(fmt.Sprintf("user requires %s scope to invite other users", authorization.ManageMembersScope))
+		}
+	}
+
+	// Iterate through all of the invitations and confirm that for each one:
+	// 1) a valid user has been specified via its User ID, e-mail address or username
+	// 2) any roles specified are valid roles for the organization, team or security group
+	// For each invitation, ensure that the IdentityID value can be found and set it
+	for _, invitation := range invitations {
+		// Load the identity
+		identity, err := s.repo.Identities().Load(ctx, *invitation.IdentityID)
+		if err != nil {
+			return errors.NewInternalError(ctx, err)
+		}
+
+		if !identity.IsUser() {
+			return errors.NewBadParameterErrorFromString("Identity ID", invitation.IdentityID, "identity is not a user")
```

Comment:
No tests for this usecase (when the identity is not a user)

Reply:
Test added